### PR TITLE
fix(ci): wire coverage observability into CI pipelines (Phase 1)

### LINKED ISSUE #1399: [FEATURE] Coverage Phase 1: Wire observability into CI
**Is your feature request related to a problem? Please describe.**
CI has no coverage measurement. Contributors cannot see coverage impact
of their changes, and regressions land silently.

**Describe the solution you'd like**
Wire `pytest-cov` into both CI pipelines (unit and integration) with
report-only output to `$GITHUB_STEP_SUMMARY`. No `fail_under` gate --
this phase establishes observability only.

### Changes

1. Add `pytest-cov` to dev dependencies
2. Add `[tool.coverage.run]` and `[tool.coverage.report]` to `pyproject.toml`
   - `source = ["src/apm_cli"]`, `branch = true`, `parallel = true`
   - Exclude patterns: `pragma: no cover`, `TYPE_CHECKING`, `NotImplementedError`
3. Update `ci.yml` unit test step with `--cov` flags
4. Add coverage summary step writing markdown table to `$GITHUB_STEP_SUMMARY`
5. Update `ci-integration.yml` sharded test step with `--cov` flags
6. Add per-shard coverage artifact upload
7. Add fan-in job to download, merge, and render integration coverage summary
8. No `fail_under` gate -- observability only

**Describe alternatives you've considered**
- Using Codecov or Coveralls -- rejected to avoid third-party dependency;
  GitHub-native `$GITHUB_STEP_SUMMARY` keeps everything in-repo.

**Additional context**
Parent: #1398 -- Phase 1 of the progressive coverage strangler-fig ratchet.

### Acceptance criteria

- [ ] Unit coverage summary appears in CI job summary (markdown table)
- [ ] Integration coverage summary appears in fan-in job summary
- [ ] Both summaries include overall % and collapsible lowest-coverage files
- [ ] No PR is blocked by coverage -- report only

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
## TL;DR

Adds test coverage reporting to both CI pipelines (unit and integration) as Phase 1 of the progressive coverage ratchet (#1398). Coverage summaries appear in GitHub Actions job summaries as markdown tables — no PR is blocked.

> [!NOTE]
> Phase 1 is observability only — no `fail_under` gates. Phases 2–4 (#1401, #1402, #1400) will progressively tighten thresholds.

## Problem (WHY)

- [x] CI has no coverage measurement today. Contributors cannot see the coverage impact of their changes, and regressions land silently (#1398).
- [!] Without coverage visibility, there is no baseline to gate against in future phases.

## Approach (WHAT)

| # | Fix |
|---|-----|
| 1 | Add `[tool.coverage.run]` and `[tool.coverage.report]` to `pyproject.toml` with `branch = true`, `parallel = true`, and standard exclusion patterns |
| 2 | Wire `--cov` + JSON report into `ci.yml` unit test step; add coverage summary step |
| 3 | Wire `--cov` into integration shard `PYTEST_EXTRA_ARGS`; upload `.coverage` data files per shard with `include-hidden-files: true` |
| 4 | Fan-in job: `coverage combine` → `coverage json` → shared summary script for correct multi-shard merge |
| 5 | Create `scripts/coverage-summary.py` — stdlib-only shared renderer used by both pipelines |
| 6 | All coverage steps use `continue-on-error: true` / `if: always()` so coverage never blocks the required check |

## Implementation (HOW)

- **`pyproject.toml`** — Adds `[tool.coverage.run]` (source, branch, parallel) and `[tool.coverage.report]` (exclusions for `pragma: no cover`, `TYPE_CHECKING`, `NotImplementedError`, `@overload`).
- **`scripts/coverage-summary.py`** — New file. Reads `coverage.json`, renders a markdown summary (overall %, metrics table, collapsible lowest-coverage files). Writes to stdout and `$GITHUB_STEP_SUMMARY`. Exits 0 on missing input (soft-fail). Stdlib only, ASCII source.
- **`.github/workflows/ci.yml`** — Adds `--cov --cov-report=term-missing:skip-covered --cov-report=json:coverage.json` to the pytest invocation. Calls the shared summary script.
- **`.github/workflows/ci-integration.yml`** — Each shard produces a `.coverage` binary data file, renamed per-shard and uploaded with `include-hidden-files: true`. Fan-in job: checkout → Python → `pip install coverage[toml]` → download with `merge-multiple: true` → `coverage combine` → `coverage json` → summary script. Coverage summary runs before shard-result gate so observability is preserved even when shards fail.

## Diagrams

Legend: Integration coverage data flow from shards to fan-in. The yellow region highlights the `coverage combine` merge path.

```mermaid
sequenceDiagram
    participant S as Shard Job 1..4
    participant A as Upload Artifact
    participant F as Fan-in Job

    S->>S: pytest --cov
    S->>S: mv .coverage .coverage.shard-N
    S->>A: upload .coverage.shard-N
    rect rgb(255, 247, 200)
        Note over F: NEW: proper coverage combine
        A-->>F: download all shards
        F->>F: coverage combine
        F->>F: coverage json
        F->>F: coverage-summary.py
    end
    F->>F: aggregate shard results
```

## Trade-offs

- **`coverage combine` over JSON merge.** `coverage combine` is the only correct way to aggregate parallel coverage data — it handles source-file deduplication and branch-level merging internally.
- **`pip install coverage[toml]` in fan-in, not full `uv sync`.** The fan-in job only runs data processing (combine + json + summary), not tests. Installing just `coverage` keeps the step fast (~5s vs ~60s).
- **Coverage steps are soft-fail (`continue-on-error`).** Trades coverage visibility for CI stability — if coverage collection breaks, the required check still passes. Acceptable for Phase 1 observability; will be revisited when gates are introduced in Phase 2+.
- **Scenario Evidence skip clause applies.** This PR changes only CI workflow files and a standalone script — no CLI behaviour change, no user-facing code paths affected. Existing test suites run green without modification.

## Benefits

1. Coverage percentage visible in every CI run's job summary — contributors see impact before review.
2. Lowest-coverage files surfaced in a collapsible table — identifies improvement targets without requiring local tooling.
3. Integration coverage correctly merged across 4 shards via `coverage combine`.
4. Shared summary script eliminates duplication — one renderer, consistent output for both pipelines.
5. Coverage never blocks a PR — soft-fail design ensures observability does not become a gate.

## Validation

<details><summary>Lint (ruff check + format)</summary>

```
$ uv run --extra dev ruff check src/ tests/
All checks passed!

$ uv run --extra dev ruff format --check src/ tests/
775 files already formatted

$ uv run --extra dev ruff check scripts/coverage-summary.py
All checks passed!
```

</details>

<details><summary>Coverage summary script — end-to-end test</summary>

```
$ uv run pytest tests/unit/test_apm_package.py --cov --cov-report=json:coverage-test.json -q
32 passed in 2.76s

$ python3 scripts/coverage-summary.py coverage-test.json --title "Test Run"
## Test Run

**Overall: 6%** (2,762/36,735 statements)

| Metric | Value |
|--------|-------|
| Statements | 36,735 |
| Covered | 2,762 |
| Missed | 33,973 |
| Coverage | 6% |
```

</details>

<details><summary>Missing-file edge case</summary>

```
$ python3 scripts/coverage-summary.py nonexistent.json
[!] coverage-summary: nonexistent.json not found -- skipping summary.
$ echo $?
0
```

</details>

## How to test

- [ ] Push this branch and observe the CI `Build & Test (Linux)` job summary — it should show a "Unit Test Coverage" markdown table with overall % and a collapsible lowest-coverage files section.
- [ ] Enqueue a merge-queue run and observe the `Integration Tests (Linux)` fan-in job summary — it should show an "Integration Test Coverage" table with correctly merged shard data.
- [ ] Verify no `fail_under` gate exists — coverage should never cause a red check.

Closes #1399

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: scripts/coverage-summary.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    lines.append(f"## {title}")
+    lines.append("")
+    lines.append(f"**Overall: {pct}%** ({covered:,}/{stmts:,} statements)")
+    lines.append("")
+    lines.append("| Metric | Value |")
+    lines.append("|--------|-------|")
+    lines.append(f"| Statements | {stmts:,} |")
+    lines.append(f"| Covered | {covered:,} |")
+    lines.append(f"| Missed | {miss:,} |")
+    lines.append(f"| Coverage | {pct}% |")
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply scripts/coverage-summary.py:56]: Good catch on clarity. Fixed in dfa0fce7 -- now reads `covered_lines` directly from the JSON totals instead of computing it. The values are identical (`num_statements - missing_lines == covered_lines` in coverage.py's JSON output), but using the explicit field is clearer.

File: scripts/coverage-summary.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        lines.append("|------|-------|------|-------|")
+        for fpath, fdata in bottom:
+            s = fdata.get("summary", {})
+            # Strip common prefix to keep the table narrow.
+            short = fpath.replace("src/apm_cli/", "")
+            fp = s.get("percent_covered_display", "?")
+            lines.append(
+                f"| `{short}` | {s.get('num_statements', 0)}"
+                f" | {s.get('missing_lines', 0)} | {fp}% |"
+            )
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply scripts/coverage-summary.py:83]: Intentional -- this is an internal CI script for the APM repo, not a general-purpose tool. The `src/apm_cli/` prefix matches the `source = ["src/apm_cli"]` in `[tool.coverage.run]` and will always be the prefix in this repo's coverage.json. Making it configurable would add complexity without benefit here.

File: .github/workflows/ci-integration.yml [hunk 3]
```diff
@@ -202,14 +202,31 @@ jobs:
...
+          if [ -f .coverage ]; then
+            mv .coverage .coverage.shard-${{ matrix.shard }}
+          fi
+
+      - name: Upload coverage data
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: integration-coverage-shard-${{ matrix.shard }}
+          path: .coverage.shard-${{ matrix.shard }}
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply .github/workflows/ci-integration.yml:225]: pytest-cov handles this automatically: when running with xdist (`-n 2`), it creates per-worker `.coverage.<hash>` files and then combines them into a single `.coverage` file at the end of the test run. The `parallel = true` config tells coverage.py to produce uniquely named data files, and pytest-cov's built-in combine step merges them before the process exits. So `.coverage` will exist after the shard completes.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -106,8 +106,15 @@ jobs:
     - name: Install dependencies
       run: uv sync --extra dev --extra build
 
-    - name: Run tests
-      run: uv run pytest tests/unit tests/test_console.py -n auto --dist worksteal
+    - name: Run tests with coverage
+      run: uv run pytest tests/unit tests/test_console.py -n auto --dist worksteal --cov --cov-report=term-missing:skip-covered --cov-report=json:coverage.json
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply .github/workflows/ci.yml:110]: pytest-cov auto-discovers `pyproject.toml` (standard behaviour since coverage.py 5.x). The `--cov` flag without a package arg uses the `source` setting from `[tool.coverage.run]` (`src/apm_cli`), so it only measures the right code. Adding `--cov-config=pyproject.toml` would be redundant -- it's the default search path.

File: .github/workflows/ci-integration.yml [hunk 5]
```diff
@@ -219,6 +236,40 @@ jobs:
...
+      # gate runs last so it can still fail the job.
+      - uses: actions/checkout@v4
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - name: Install coverage
+        run: pip install "coverage[toml]"
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply .github/workflows/ci-integration.yml:250]: Agreed this matters for gating phases. For Phase 1 (observability only, no `fail_under`), an unpinned install is acceptable -- a coverage version bump cannot break CI since we never fail on the number. We'll pin when gates are introduced in Phase 2 (#1401).

**Review by danielmeppiel — APPROVED** [review]
